File: Python/benchmark_v2.py
import time
import random
from RSPCVT import RSPCVT

def tranthong_interruptible(xstart, ystart, xend, yend, callback):
    x, y = xstart, ystart
    
    deltax = abs(xend - xstart)
    signdx = 1 if xend >= xstart else -1
    
    deltay = abs(yend - ystart)
    signdy = 1 if yend >= ystart else -1

    # If callback returns True, stop the ray immediately
    if callback(x, y):
        return

    test = -1 if signdy == -1 else 0

    if deltax >= deltay:
        test = (deltax + test) >> 1
        for _ in range(deltax):
            test -= deltay
            x += signdx
            if test < 0:
                y += signdy
                test += deltax
            if callback(x, y):
                break
    else:
        test = (deltay + test) >> 1
        for _ in range(deltay):
            test -= deltax
            y += signdy
            if test < 0:
                x += signdx
                test += deltay
            if callback(x, y):
                break

def create_random_map(size, density):
    return [[random.random() < density for _ in range(size)] for _ in range(size)]

def run_fair_benchmark():
    ITERATIONS = 10000
    RADIUS = 20
    MAP_SIZE = 60
    DENSITY = 0.25
    SEED = 42

    fov_engine = RSPCVT(RADIUS)
    
    # Pre-calculate targets
    targets = []
    for i in range(-RADIUS, RADIUS + 1):
        for j in range(-RADIUS, RADIUS + 1):
            if i*i + j*j <= RADIUS*RADIUS:
                targets.append((i, j))

    print(f"Benchmarking Python (Fair): Radius {RADIUS}, {ITERATIONS} iterations")
    print("-" * 50)

    # 1. Map Gen Overhead
    random.seed(SEED)
    start_map = time.perf_counter()
    for _ in range(ITERATIONS):
        create_random_map(MAP_SIZE, DENSITY)
    map_gen_total = time.perf_counter() - start_map

    # 2. Naive with Early Exit
    random.seed(SEED)
    start_naive = time.perf_counter()
    for _ in range(ITERATIONS):
        world_map = create_random_map(MAP_SIZE, DENSITY)
        ox, oy = 30, 30
        
        for tx_off, ty_off in targets:
            tx, ty = ox + tx_off, oy + ty_off
            # Passing a lambda that returns True if it hits a wall
            tranthong_interruptible(ox, oy, tx, ty, lambda x, y: world_map[y][x] if 0 <= x < MAP_SIZE and 0 <= y < MAP_SIZE else True)
    
    naive_total = (time.perf_counter() - start_naive) - map_gen_total

    # 3. RSPCVT
    random.seed(SEED)
    start_trie = time.perf_counter()
    for _ in range(ITERATIONS):
        world_map = create_random_map(MAP_SIZE, DENSITY)
        ox, oy = 30, 30
        
        fov_engine.fov(ox, oy, 
            lambda x, y: world_map[y][x] if 0 <= x < MAP_SIZE and 0 <= y < MAP_SIZE else True,
            lambda x, y: None
        )

    trie_total = (time.perf_counter() - start_trie) - map_gen_total

    # Results
    print(f"Avg Map Gen Time:    {map_gen_total / ITERATIONS:.6f} s")
    print(f"Total Naive Time:    {naive_total:.4f} s (excl. map gen)")
    print(f"Total RSPCVT Time:   {trie_total:.4f} s (excl. map gen)")
    print("-" * 50)
    print(f"Speedup Factor:      {naive_total / trie_total:.2f}x faster")
    print(f"Avg RSPCVT FOV:      {trie_total / ITERATIONS:.6f} s")

if __name__ == "__main__":
    run_fair_benchmark()

File: Lua/benchmark_v2.lua
local bit = require 'bit'

local function Difference(xstrt, xnd)
  if xnd >= xstrt then
    return xnd - xstrt, 1
  else
    return xstrt - xnd, -1
  end
end

local function TranThongInterruptible(xstart, ystart, xend, yend, callback)
  local x = xstart
  local y = ystart

  local deltax, signdx = Difference(xstart, xend)
  local deltay, signdy = Difference(ystart, yend)

  -- If the start tile blocks, we stop immediately
  if callback(x, y) then return end

  local test = signdy == -1 and -1 or 0

  if deltax >= deltay then
    test = bit.rshift(deltax + test, 1)
    for i = 1, deltax do
      test = test - deltay
      x = x + signdx
      if test < 0 then
        y = y + signdy
        test = test + deltax
      end
      if callback(x, y) then break end
    end
  else
    test = bit.rshift(deltay + test, 1)
    for i = 1, deltay do
      test = test - deltax
      y = y + signdy
      if test < 0 then
        x = x + signdx
        test = test + deltay
      end
      if callback(x, y) then break end
    end
  end
end

local RSPCVT = require 'RSPCVT'
local TranThong = TranThongInterruptible
local getTime = os.clock

local ITERATIONS = 50000
local RADIUS = 20
local MAP_SIZE = 60
local DENSITY = 0.25

local fov_engine = RSPCVT(RADIUS)

-- Pre-calculate target list
local targets = {}
for i = -RADIUS, RADIUS do
  for j = -RADIUS, RADIUS do
    if i*i + j*j <= RADIUS*RADIUS then
      table.insert(targets, {x = i, y = j})
    end
  end
end

local function create_random_map(size, density)
  local map = {}
  for y = 0, size do
    map[y] = {}
    for x = 0, size do
      map[y][x] = math.random() < density
    end
  end
  return map
end

local function run_benchmark()
  print(string.format("Benchmarking (Fair): Radius %d, %d iterations", RADIUS, ITERATIONS))
  print("--------------------------------------------------")

  -- 1. Baseline Map Gen
  math.randomseed(42)
  local start_map = getTime()
  for _ = 1, ITERATIONS do
    create_random_map(MAP_SIZE, DENSITY)
  end
  local map_gen_total = getTime() - start_map

  -- 2. Naive with Early Exit
  math.randomseed(42)
  local start_naive = getTime()
  for _ = 1, ITERATIONS do
    local map = create_random_map(MAP_SIZE, DENSITY)
    local ox, oy = 30, 30
    
    for k = 1, #targets do
      local t = targets[k]
      TranThong(ox, oy, ox + t.x, oy + t.y, function(x, y)
        return map[y] and map[y][x] -- Return true to stop the ray
      end)
    end
  end
  local naive_total = (getTime() - start_naive) - map_gen_total
  
  -- 3. RSPCVT
  math.randomseed(42)
  local start_trie = getTime()
  for _ = 1, ITERATIONS do
    local map = create_random_map(MAP_SIZE, DENSITY)
    local ox, oy = 30, 30
    
    fov_engine:FOV(ox, oy, 
      function(x, y) return map[y] and map[y][x] end,
      function(x, y) end
    )
  end
  local trie_total = (getTime() - start_trie) - map_gen_total

  -- Output
  print(string.format("Avg Map Gen Time:    %.6f s", map_gen_total / ITERATIONS))
  print(string.format("Total Naive Time:    %.4f s (excl. map gen)", naive_total))
  print(string.format("Total RSPCVT Time:   %.4f s (excl. map gen)", trie_total))
  print("--------------------------------------------------")
  print(string.format("Speedup Factor:      %.2fx faster", naive_total / trie_total))
  print(string.format("Avg RSPCVT FOV:      %.6f s", trie_total / ITERATIONS))
end

run_benchmark()